### PR TITLE
Add Support For Specifying Managed Image by ID

### PR DESCRIPTION
**Pull request by dave-tucker**
We use [Packer](https://packer.io) for building our base images for use in Jenkins.
When following the [Azure instructions for using Packer](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/build-image-with-packer) the result is an Image and therefore couldn't be used in this plugin as the only supported Custom Image requires a VHD. The [Packer documentation](https://www.packer.io/docs/builders/azure.html) also states:

> If you are creating a VHD, you must start with a VHD. Likewise, if you want to create a managed image you must start with a managed image. When creating a VHD the following two options are required.

So it seems like it's not easy to get a VHD from a customised Managed Image 😭 

To use images created with Packer I've added a new `Image ID` field to the `Image Reference` section:
image-config

I've updated the help text appropriately and made sure the validation only allows for either the Image Id *or* a complete set of the other parameters.

I couldn't figure out how to re-use the existing JSON templates to supply either the `image_id` or the `publisher, `offer` etc... so I created two new ones.

So far, I've tested this in our test environment and I'm seeing deployment time half compared to installing software in the init script! :sunglasses:

### CODE REVIEW

**Comment by zackliu** [comment]
One concern here. Why don't we use a separate `ImageReferenceType` named like "Customer Managed Image" instead of put configuration in the `Image Reference`. It will be much clearer for the users and more convenient for type judgment.

You judge the type between "Custom Image" or "Marketplace Image" by checking whether they filled in the blank. Instead we can add an image reference type and add a parameter named `ImageReferenceType` in the constructor of `AzureVMAgentTemplate`. Then you can get the accurate type.

You may notice we get `ImageReferenceType` in the function `getImageReferenceType()` by guessing. It is the legacy code and actually after we used the new UI library, we can get what we've chosen in dropdownlist through a parameter in the construct which has the same name with dropdownList.

Except this, other change looks great.

**Comment by dave-tucker** [comment]
Thanks @zackliu. Using a new type would be a lot better! I'll take a look at that today.

**Comment by w32-blaster** [comment]
Any ETA on when will be this released?

**Comment by dave-tucker** [comment]
I'm still working on my iatcg, but I'm hoping to have it ready early this week... Then it's up to the maintainers when to add this to a release

**Comment by zackliu** [comment]
@dave-tucker If you finished your PR, we will release it ASAP.

**Comment by dave-tucker** [comment]
Oh, yes and I also added some proper validation to check that the Id is valid 😀

File: src/main/java/com/microsoft/azure/vmagent/builders/AdvancedImageFluent.java [hunk 1]
```diff
@@ -65,11 +67,13 @@ public T withCustomImage(String imageUrl) {
...
     }
 
-    public T withReferenceImage(String imagePublisher,
+    public T withReferenceImage(String imageId,
+                                String imagePublisher,
                                 String imageOffer,
                                 String imageSku,
                                 String imageVersion) {
         this.imageReferenceType = "reference";
+        this.imageId = imageId;
```
Comment (zackliu) [inline src/main/java/com/microsoft/azure/vmagent/builders/AdvancedImageFluent.java]: `imageId` should not in the `withReferenceImage()`. Please add a new function maybe `withCustomManagedImage()` and  set `this.imageReferenceType = "customImage"`
And thus you need to do some change in `AdvancedImageBuilder`.

File: src/main/java/com/microsoft/azure/vmagent/builders/AdvancedImageBuilder.java [hunk 2]
```diff
@@ -12,7 +12,8 @@ public AdvancedImageBuilder(AdvancedImageFluent<?> fluent) {
     public AdvancedImageBuilder(AdvancedImageFluent<?> fluent, AdvancedImage image) {
         this.fluent = fluent;
         fluent.withCustomImage(image.getImage());
-        fluent.withReferenceImage(image.getImagePublisher(),
+        fluent.withReferenceImage(image.getImageId(),
```
Comment (zackliu) [inline src/main/java/com/microsoft/azure/vmagent/builders/AdvancedImageBuilder.java]: Seems here is a left bug. We should justify the `imageReferenceType` first and then use ` fluent.withCustomImage()`, `fluent.withReferenceImage()` or `withCustomManagedImage()`

**Comment by dave-tucker** [comment]
@zackliu thanks for the review comments. They should now be addressed.

**Pull request merged**